VisualScriptFunctionNodes Improvements

Gallilus opened the pull request:
This PR improves and streamlines the workflow for VisualScriptFunctionNodes. Call Set Get
Removes redundant method_select popup
And Fix issues with drop offset and instancing Set Get

<!--
Pull requests should always be made for the `master` branch first, as that's
where development happens and the source of all future stable release branches.

Relevant fixes are cherry-picked for stable branches as needed.

Do not create a pull request for stable branches unless the change is already
available in the `master` branch and it cannot be easily cherry-picked.
Alternatively, if the change is only relevant for that branch (e.g. rendering
fixes for the 3.2 branch).
-->
Supersedes #49184
Supersedes #49172
Fixes #26717
Fixes #49146
Possible Fix #37053

### Editor changes
* Fix issue when of dropposition when in different zoom levels.
* Fix drag in Set and Get.
### VisualScriptFuncNode's changes
* Harmonize and improve the visual representation of the VisualScriptFuncNode's and there different modes.
* Update Functionality of VisualScriptFunctionGet to include pass when using instance mode.

[example project.zip](https://github.com/godotengine/godot/files/6695328/example.project.zip)
[example script.zip](https://github.com/godotengine/godot/files/6695334/example.script.zip)

<img width="1198" alt="screeshot" src="https://user-images.githubusercontent.com/20573784/122947038-22e78080-d37a-11eb-9105-c6d9877d1537.PNG">
caption = what
text (sequence) = where / object info
I-O's = details 

# Original
![image](https://user-images.githubusercontent.com/20573784/122654788-0a931e00-d14e-11eb-9323-f9c2e1c3fa31.png)

YeldhamDev commented:
Small typo in the commit message: "Imrovements"

charlesmlamb commented:
ok, 10 / 10, superb . .

fire commented:
Reviewing.

fire commented on modules/visual_script/visual_script_editor.cpp:
```diff
@@ -3235,18 +3166,62 @@ void VisualScriptEditor::_selected_connect_node(const String &p_text, const Stri
 		vnode = n;
 	}
+	source_path = String();
+	source_node = NULL;
```
Minor:

`NULL` usage is inconsistent with `nullptr`.

fire commented on modules/visual_script/visual_script_editor.h:
```diff
@@ -179,6 +179,9 @@ class VisualScriptEditor : public ScriptEditorBase {
 	void connect_data(Ref<VisualScriptNode> vnode_old, Ref<VisualScriptNode> vnode, int new_id);
 
+	NodePath source_path;
+	Node *source_node;
```
Init to `nullptr`.

fire reviewed (approved):
Improving the situation.

1. Reviewed the code.
2. Reviewed the example photo
3. Still need test with demo projects to see the visual scripting's effect.

Please try further QA and resolve the style changes.

Approved on the design.

Note:
Trying to create a get_position on Node2D crashes. Cannot reproduce again.

DavidCambre commented:
@fire, sorry I'm confused.
Please try further QA and resolve the style changes. Is this directed at me? 
Do I need to make it with nullptr's? Do you have a example?

Gallilus committed:
Improve and streamline VisualScriptFuncNodes Call Set Get

This PR improves and streamlines the workflow for VisualScriptFunctionNodes Call Set Get
Uniform design.
Drag in set-get from tree is now working.
Removes redundant method_select popup.

DavidCambre commented:
@fire
Tested even got a few small extras that where overlooked.
Set Basic Type now has sequence ports.
Ready for merge

charlesmlamb commented:
Thanks, for this, every improvement to Visual script matters to me, Godot 4 will be amazing . .

DavidCambre commented:
Do I need a new approval before it is possible to merge?
I'm dreaming of getting this to 3.x before 3.3.3

Calinou commented:
> I'm dreaming of getting this to 3.x before 3.3.3

3.3.x will only get bug fixes; new features will be available in 3.4.

akien-mga commented on modules/visual_script/visual_script_func_nodes.cpp:
```diff
@@ -1036,17 +1045,18 @@ String VisualScriptPropertySet::get_caption() const {
 }
 
 String VisualScriptPropertySet::get_text() const {
+	if (!has_input_sequence_port())
+		return "";
```
```suggestion
	if (!has_input_sequence_port()) {
		return "";
	}
```

akien-mga reviewed (approved):
Assign from a small style issue, this seems pretty solid. I'm not very familiar with VisualScript but it seems to improve UX, and we can iterate further from there.

akien-mga commented:
Once merged, feel free to make a PR to backport the changes to `3.x` for inclusion in 3.4. But please also make sure that the changes do not break compatibility for existing scripts (I think not, but again I don't know much about VisualScript).

akien-mga commented:
Thanks!

DavidCambre commented:
Real time support of Remi is Awesome!!
